substrate/client/rpc-spec-v2/src/archive/archive_storage.rs

```diff
@@ -127,3 +140,738 @@ where
+				let Some(value) = value else {
+					return Ok(None);
+				};
+
+				let hash = self.client.query_hash(hash, &key, maybe_child_trie.as_ref())?;
+				let Some(hash) = hash else {
+					return Ok(None);
+				};
+
+				Ok(Some(FetchedStorage::Both { value, hash }))
+			},
+		}
+	}
+
+	/// Check if the key belongs to the provided query items.
+	///
+	/// A key belongs to the query items when:
+	/// - the provided key is a prefix of the key in the query items.
+	/// - the query items are empty.
+	///
+	/// Returns an optional `FetchStorageType` based on the query items.
+	/// If the key does not belong to the query items, returns `None`.
+	fn belongs_to_query(key: &StorageKey, items: &[DiffDetails]) -> Option<FetchStorageType> {
+		// User has requested all keys, by default this fallbacks to fetching the value.
+		if items.is_empty() {
+			return Some(FetchStorageType::Value)
+		}
+
+		let mut value = false;
+		let mut hash = false;
+
+		for item in items {
+			if key.as_ref().starts_with(&item.key.as_ref()) {
+				match item.return_type {
+					ArchiveStorageDiffType::Value => value = true,
+					ArchiveStorageDiffType::Hash => hash = true,
+				}
+			}
+		}
+
+		match (value, hash) {
+			(true, true) => Some(FetchStorageType::Both),
+			(true, false) => Some(FetchStorageType::Value),
+			(false, true) => Some(FetchStorageType::Hash),
+			(false, false) => None,
+		}
+	}
+
+	/// Send the provided result to the `tx` sender.
+	///
+	/// Returns `false` if the sender has been closed.
+	fn send_result(
+		tx: &mpsc::Sender<ArchiveStorageDiffEvent>,
+		result: FetchedStorage,
+		operation_type: ArchiveStorageDiffOperationType,
+		child_trie_key: Option<String>,
+	) -> bool {
+		let items = match result {
+			FetchedStorage::Value(storage_result) | FetchedStorage::Hash(storage_result) =>
+				vec![storage_result],
+			FetchedStorage::Both { value, hash } => vec![value, hash],
+		};
+
+		for item in items {
+			let res = ArchiveStorageDiffEvent::StorageDiff(ArchiveStorageDiffResult {
+				key: item.key,
+				result: item.result,
+				operation_type,
+				child_trie_key: child_trie_key.clone(),
+			});
+			if tx.blocking_send(res).is_err() {
+				return false
+			}
+		}
+
+		true
+	}
+
+	fn handle_trie_queries_inner(
+		&self,
+		hash: Block::Hash,
+		previous_hash: Block::Hash,
+		items: Vec<DiffDetails>,
+		tx: &mpsc::Sender<ArchiveStorageDiffEvent>,
+	) -> Result<(), String> {
+		// Parse the child trie key as `ChildInfo` and `String`.
+		let maybe_child_trie = items.first().and_then(|item| item.child_trie_key.clone());
+		let maybe_child_trie_str =
+			items.first().and_then(|item| item.child_trie_key_string.clone());
+
+		// Iterator over the current block and previous block
+		// at the same time to compare the keys. This approach effectively
+		// leverages backpressure to avoid memory consumption.
+		let mut keys_iter = self.client.raw_keys_iter(hash, maybe_child_trie.clone())?;
```

Comment:
From my understanding this algorithm relies on that the keys are sorted in lexicographical order to make such comparisons but neat.

Wouldn't hurt with some unit tests for lexicographical order stuff but looks good to me

Reply:
The add/delete/modify code below looks good to me too. One option if you wanted to easily unit test this bit would be to make a separate general-purpose lexicographic_diff function something like:

```rust
fn lexicographic_diff<T, As, Bs>(left: As, right: Bs) -> impl Iterator<Item = Diff<&T>> 
where 
    As: Iterator<Item = &T>,
    Bs: Iterator<Item = &T>
{
    let mut a = left.next();
    let mut b = right.next();

    loop {
        // same code for comparing; returns Diff<T> to denote difference
    }
}

enum Diff<T> {
    Added(T),
    Deleted(T),
    Equal(T)
}
```

and then here:

```rust
for diff in lexicographic_diff(previous_keys_iter, keys_iter) {
    match diff {
        Diff::Added(key) => (ArchiveStorageDiffOperationType::Added, key),
        // etc
    }
}
```

Then the diff logic can be unit tested nice and easy :)

Reply:
That simplifies our lives quite a bit, have added the `lexicographic_diff` method 🙏